### PR TITLE
bpo-27071: rename assertCountEqual to assertPermutation

### PR DESCRIPTION
**Pull request by graingert**
https://bugs.python.org/issue27071

### CODE REVIEW

File: Doc/library/unittest.rst [hunk 1]
```diff
@@ -1212,19 +1213,20 @@ Test cases
...
-      regardless of their order. When they don't, an error message listing the
-      differences between the sequences will be generated.
+      Test that sequence *first* is a permutation of sequence *second*:
+      If not, an error message listing the differences between
+      the sequences will be generated.
 
       Duplicate elements are *not* ignored when comparing *first* and
       *second*. It verifies whether each element has the same count in both
       sequences. Equivalent to:
       ``assertEqual(Counter(list(first)), Counter(list(second)))``
```
Comment (graingert) [inline Doc/library/unittest.rst:1226]: what do we think about

```rst
Equivalent to:
``assertIn(tuple(first), itertools.permutations(second))``
but much more efficiently
```

**Comment by graingert** [comment]
@voidspace ^

File: Doc/library/unittest.rst (hunk 1 above)
Comment (graingert) [reply Doc/library/unittest.rst:1226]: @voidspace ^

File: Doc/library/unittest.rst [hunk 2]
```diff
@@ -1212,19 +1213,20 @@ Test cases
...
       but works with sequences of unhashable objects as well.
 
-      .. versionadded:: 3.2
+      .. versionadded:: 3.9
```
Comment (graingert) [inline Doc/library/unittest.rst:1229]: something like around here: 
```rst
In Python 3.2 to 3.8 this method is named assertCountEqual.

.. versionadded:: 3.9
```

**Review by terryjreedy — DISMISSED** [review]
REMOVE ALL CHANGES TO idlelib/idle-test/*.py.
See msg 352648 on the issue.

**Comment by terryjreedy** [comment]
Tests fail because aliasing somehow did not work.
```
ERROR: testAssertPermutation (unittest.test.test_case.Test_TestCase)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "d:\a\1\s\lib\unittest\test\test_case.py", line 995, in testAssertPermutation
    self.assertEqual(self.assertCountEqual, self.assertPermutation)
AttributeError: 'Test_TestCase' object has no attribute 'assertCountEqual'
```
Besides which, the assertion should be `assertIs`.

File: Lib/unittest/case.py [hunk 3]
```diff
@@ -1254,6 +1253,8 @@ def assertCountEqual(self, first, second, msg=None):
             msg = self._formatMessage(msg, standardMsg)
             self.fail(msg)
 
+    assertCountEquals = assertPermutation
```
Comment (graingert) [inline Lib/unittest/case.py]:
```suggestion
    assertCountEqual = assertPermutation
```

**Comment by graingert** [comment]
@terryjreedy
> Tests fail because aliasing somehow did not work.
> 
> ```
> ERROR: testAssertPermutation (unittest.test.test_case.Test_TestCase)
> ----------------------------------------------------------------------
> Traceback (most recent call last):
>   File "d:\a\1\s\lib\unittest\test\test_case.py", line 995, in testAssertPermutation
>     self.assertEqual(self.assertCountEqual, self.assertPermutation)
> AttributeError: 'Test_TestCase' object has no attribute 'assertCountEqual'
> ```
It was a typo in my alias
> 
> Besides which, the assertion should be `assertIs`.
no it has to be assertEqual because methods do not compare `is`:

```python
import unittest


class Ham:
    def spam(self):
        pass

    eggs = spam

    def bacon(self):
        pass


class TestHam(unittest.TestCase):
    def test_eggs_is_eggs(self):
        h = Ham()
        self.assertIs(h.eggs, h.eggs)

    def test_spam_is_spam(self):
        h = Ham()
        self.assertIs(h.spam, h.spam)

    def test_spam_is_eggs(self):
        h = Ham()
        self.assertIs(h.spam, h.eggs)

    def test_eggs_equal_eggs(self):
        h = Ham()
        self.assertEqual(h.eggs, h.eggs)

    def test_spam_equal_spam(self):
        h = Ham()
        self.assertEqual(h.spam, h.spam)

    def test_spam_equal_eggs(self):
        h = Ham()
        self.assertEqual(h.spam, h.eggs)

    def test_spam_equal_bacon(self):
        h = Ham()
        self.assertEqual(h.spam, h.bacon)


if __name__ == "__main__":
    unittest.main()
```

```
test_eggs_equal_eggs (__main__.TestHam) ... ok
test_eggs_is_eggs (__main__.TestHam) ... FAIL
test_spam_equal_bacon (__main__.TestHam) ... FAIL
test_spam_equal_eggs (__main__.TestHam) ... ok
test_spam_equal_spam (__main__.TestHam) ... ok
test_spam_is_eggs (__main__.TestHam) ... FAIL
test_spam_is_spam (__main__.TestHam) ... FAIL

======================================================================
FAIL: test_eggs_is_eggs (__main__.TestHam)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "foo.py", line 17, in test_eggs_is_eggs
    self.assertIs(h.eggs, h.eggs)
AssertionError: <bound method Ham.spam of <__main__.Ham object at 0x7f85b504feb0>> is not <bound method Ham.spam of <__main__.Ham object at 0x7f85b504feb0>>

======================================================================
FAIL: test_spam_equal_bacon (__main__.TestHam)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "foo.py", line 41, in test_spam_equal_bacon
    self.assertEqual(h.spam, h.bacon)
AssertionError: <bound method Ham.spam of <__main__.Ham object at 0x7f85b504ff70>> != <bound method Ham.bacon of <__main__.Ham object at 0x7f85b504ff70>>

======================================================================
FAIL: test_spam_is_eggs (__main__.TestHam)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "foo.py", line 25, in test_spam_is_eggs
    self.assertIs(h.spam, h.eggs)
AssertionError: <bound method Ham.spam of <__main__.Ham object at 0x7f85b504feb0>> is not <bound method Ham.spam of <__main__.Ham object at 0x7f85b504feb0>>

======================================================================
FAIL: test_spam_is_spam (__main__.TestHam)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "foo.py", line 21, in test_spam_is_spam
    self.assertIs(h.spam, h.spam)
AssertionError: <bound method Ham.spam of <__main__.Ham object at 0x7f85b504ff70>> is not <bound method Ham.spam of <__main__.Ham object at 0x7f85b504ff70>>

----------------------------------------------------------------------
Ran 7 tests in 0.001s

FAILED (failures=4)
```

**Comment by graingert** [comment]
I have made the requested changes; please review again

**Comment by csabella** [comment]
This was rejected on the bug tracker, so I am going to close the PR.

**Comment by tadeo** [comment]
Hi, I know this name issue turned controversial but everyone agrees assertCountEqual is not a good fit.
I just want to drop something different to motivate a new different perspective.
`assertSameItems`
`assertShareItems`
Just think about new programmers approaching Python: this name needs to be simple and accessible for everyone.
Python deserves to get this stone in the shoe removed.
Thank you!

**Comment by graingert** [comment]
> Hi, I know this name issue turned controversial but everyone agrees assertCountEqual is not a good fit.
> I just want to drop something different to motivate a new different perspective.
> `assertSameItems`
> `assertShareItems`
> Just think about new programmers approaching Python: this name needs to be simple and accessible for everyone.
> Python deserves to get this stone in the shoe removed.
> Thank you!

sameItems doesn't work because two lists can have the same items as each other, but not be a permutation of each other eg: [egg, egg, ham] and [egg, ham]

**Comment by tadeo** [comment]
@graingert The first sentence in the [official documentation](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertCountEqual) defines it using "same elements":
> Test that sequence _first_ contains the same elements as _second_, regardless of their order.
So it seems `assertSameElements` can be an improvement, and `assertUnorderedSameItems` or `assertUnorderedSameElements` being longer but preciser.
Alternatively, after your answer I guess `assertPermutation` makes a good candidate for simplicity.

**Comment by graingert** [comment]
The documentation is also wrong here: "same elements" is wrong. Two collections can have the same elements but not be a permutation of each other

**Pull request closed without merging**